### PR TITLE
Refactor FooterHelper

### PR DESCRIPTION
**Pull request by JoeCohen**
- Raises `FooterHelper` test coverage to 100%
- Reduces `FooterHelper#show_object_footer` AbcSize from  82.84 to <= 17 (default)
- Reduces `FooterHelper#show_authors_and_editors` AbcSize from  39.57 to <= 17 (default)
- Fixes all other Rubocop offenses in FooterHelper

### CODE REVIEW

File: app/helpers/footer_helper.rb [hunk 1]
```diff
@@ -82,61 +113,102 @@ def show_authors_and_editors(obj)
...
+      [:footer_created_by.t(user: user_link(obj.user),
+                            date: obj.created_at.web_time)]
+    else
+      []
+    end
+  end
+
+  def html_for_non_versioned_object(obj)
+    html = []
+    if obj.created_at
```
Comment (JoeCohen) [inline app/helpers/footer_helper.rb:190]: Can we avoid this conditional? 
See https://github.com/MushroomObserver/mushroom-observer/pull/1510/files#r1197958451

File: app/helpers/footer_helper.rb [hunk 2]
```diff
@@ -82,61 +113,102 @@ def show_authors_and_editors(obj)
...
+  end
+
+  def html_for_latest_version(obj)
+    latest_user = User.safe_find(obj.versions.latest.user_id)
+    html = html_created_by(obj)
+
+    if latest_user && obj.updated_at
+      html << :footer_last_updated_by.t(user: user_link(latest_user),
+                                        date: obj.updated_at.web_time)
+    elsif obj.updated_at
```
Comment (JoeCohen) [inline app/helpers/footer_helper.rb:172]: Can we avoid the `updated_at` conditionals?

File: app/helpers/footer_helper.rb [hunk 3]
```diff
@@ -82,61 +113,102 @@ def show_authors_and_editors(obj)
...
+      html << :footer_last_updated_at.t(date: obj.updated_at.web_time)
+    end
+
+    html
+  end
+
+  def html_created_by(obj)
+    if obj.created_at
```
[the remark this answers is not in the document]
Comment (nimmolo) [reply app/helpers/footer_helper.rb:180]: @JoeCohen IIRC there used to be objects without a `created_at`, but that may not be the case anymore, and probably should not be. I might have it confused with `updated_at`, though. 

I have changes coming to the HTML generated in `show_object_footer` in Boostrap 5, but i can adapt them to whatever refactor you do here.

File: app/helpers/footer_helper.rb [hunk 4]
```diff
@@ -82,61 +113,102 @@ def show_authors_and_editors(obj)
...
+  end
+
+  ###############################################################
+
+  private
 
-    # Latest version of non-versioned object.
+  def html_for_old_version_of_versioned_object(obj, num_versions)
+    html = [:footer_version_out_of.t(num: obj.version, total: num_versions)]
+    return html unless obj.updated_at
```
[the remark this answers is not in the document]
Comment (nimmolo) [reply app/helpers/footer_helper.rb:135]: @JoeCohen I'd say just double-check against `schema.rb` to be sure the common-sense assumption is actually true!

**Comment by JoeCohen** [comment]
Interesting! Thanks for that info.
I'll see if there are showable objects that lack created_at and/or
updated_at.

**Comment by JoeCohen** [comment]
18054/181904 Comments have `updated_at: nil`. The most recent of these is from 2010.
- We don't use footers with Comments. So this should have no effect on this PR.
- @pellaea: Should we fix these? (I think so.)

**Comment by pellaea** [comment]
My two cents: even if these should always be non-nil, I would be careful, because the consequence of being wrong is it will give a 500 error.  It used to be harmless to add sanity checks.  I assume that the problem is  nowadays such sanity checks come with the baggage of needing test cases.  But since these cases should never occur it is therefore impossible to test.  Does that about sum it up?  Clearly the easiest solution is to get rid of the check.  And if we don't actually care if MO throws 500 errors from time to time (and we get good at ignoring the cronjob email that tells us about these errors), then it's a win all around.  Unfortunately, if we do actually care about minimizing the frequency of 500 errors, we would need to track down and fix whatever resulted in these impossible conditions actually occurring.  But that is generally extremely challenging to do.  What are the trade-offs in time spent writing tests versus tracking down pathological events?  (Versus lowering our expectations of MO! :)

**Comment by JoeCohen** [comment]
Thanks Jason!
I'll keep it as-is (with the checks for the supposedly
impossible conditions).

**Pull request merged**